python/graphframes/examples.py

```diff
@@ -57,3 +60,216 @@ def friends(self):
+    Thin Junction Trees." AISTATS, 2011.
+
+    The BP algorithm works by:
+
+    * Coloring the graph by assigning a color to each vertex such that no neighboring vertices
+      share the same color.
+    * In each step of BP, update all vertices of a single color.  Alternate colors.
+     """
+    @classmethod
+    def main(cls):
+        """Run the algorithm for an example problem."""
+        conf = SparkConf().setAppName("BeliefPropagation example")
+        sc = SparkContext.getOrCreate(conf)
+        sql = SQLContext.getOrCreate(sc)
+
+        # create graphical model g of size 3 x 3
+        g = Graphs(sql).gridIsingModel(3)
+        print("Original Ising model:")
+        g.vertices.show()
+        g.edges.show()
+
+        # run BP for 5 iterations
+        numIter = 5
+        results = cls.runBPwithGraphFrames(g, numIter)
+
+        # display beliefs
+        beliefs = results.vertices.select('id', 'belief')
+        print("Done with BP. Final beliefs after {} iterations:".format(numIter))
+        beliefs.show()
+
+        sc.stop()
+
+    @staticmethod
+    def colorGraph(g):
+        """Given a GraphFrame, choose colors for each vertex.
+
+        No neighboring vertices will share the same color. The number of colors is minimized.
+
+        This is written specifically for grid graphs. For non-grid graphs, it should be generalized,
+        such as by using a greedy coloring scheme.
+
+        :param g: Grid graph generated by :meth:`Graphs.gridIsingModel()`
+        :return: Same graph, but with a new vertex column "color" of type Int (0 or 1)
+
+        """
+
+        colorUDF = sqlFuncs.udf(
+            lambda i, j: 0 if (i + j) % 2 == 0 else 1,
+            returnType=types.IntegerType())
+        v = g.vertices.withColumn('color', colorUDF(sqlFuncs.col('i'), sqlFuncs.col('j')))
+        return GraphFrame(v, g.edges)
+
+    @classmethod
+    def runBPwithGraphFrames(cls, g, numIter):
+        """Run Belief Propagation using GraphFrame.
+
+        This implementation of BP shows how to use GraphFrame's aggregateMessages method.
+        """
+        # choose colors for vertices for BP scheduling
+        colorG = cls.colorGraph(g)
+        numColors = colorG.vertices.select('color').distinct().count()
+
+        # TODO: handle vertices without any edges
+
+        # initialize vertex beliefs at 0.0
+        gx = GraphFrame(colorG.vertices.withColumn('belief', sqlFuncs.lit(0.0)), colorG.edges)
+
+        # run BP for numIter iterations
+        for iter_ in range(numIter):
+            # for each color, have that color receive messages from neighbors
+            for color in range(numColors):
+                # Send messages to vertices of the current color.
+                # We may send to source or destination since edges are treated as undirected.
+                msgForSrc = sqlFuncs.when(
+                    AM.src()['color'] == color,
+                    AM.edge()['b'] * AM.dst()['belief'])
```

Comment:
In scala, I believe, these return new Column instances in every access. It is better to do the same in Python also. This avoids accidentally messing up with it by modifying the obtained Column object in-place.

BTW. I found out that building the python API doc gets complicated if there are such class or module level attributes that rely on the graphframes java api. The current build setup fails here as the import fails. I found this when I tried to make `GraphFrame.ID` etc. into class attributes.

Bellow are some candidates; whould some of these be good?

### 1)
Called by `AM().src`
```python
class AggregateMessages(object):
    def __init__(self):
        self._src_col_name = _java_api(SparkContext).SRC()
    @property
    def src(self):
        return sqlfunctions.col(self._src_col_name)
```
Simple alternative to the current `AM.src()`

### 2)
Called by `AM.src`
```python
class _ClassProperty(property):
    def __get__(self, cls, owner):
        return self.fget.__get__(None, owner)()

class AggregateMessages(object):
    @_ClassProperty
    @staticmethod
    def src(self):
        return sqlfunctions.col(_java_api(SparkContext).SRC())
```
No parenthesis needed, this is the one I spoke of earlier

### 3)
Called by `AM.src`
```python
class _AggregateMessages(object):
    @property
    def src(self):
        return sqlfunctions.col(_java_api(SparkContext).SRC())

AggregateMessages = _AggregateMessages()
```
No parenthesis needed, maybe controversial http://stackoverflow.com/a/6760726

### 4)
Called by `AM.src`
```python
def singleton_object(cls):
    """Class decorator that transforms (and replaces) a class definition (which
    must have a Singleton metaclass) with the actual singleton object. Ensures
    that the resulting object can still be "instantiated" (i.e., called),
    returning the same object. Also ensures the object can be pickled, is
    hashable, and has the correct string representation (the name of the
    singleton)
    """
    assert isinstance(cls, _Singleton), \
        cls.__name__ + " must use Singleton metaclass"

    def self_instantiate(self):
        return self

    cls.__call__ = self_instantiate
    cls.__hash__ = lambda self: hash(cls)
    cls.__repr__ = lambda self: cls.__name__
    cls.__reduce__ = lambda self: cls.__name__
    obj = cls()
    obj.__name__ = cls.__name__
    return obj

class _Singleton(type):
    """ A metaclass that creates a Singleton base class when called. """
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(_Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]

class Singleton(_Singleton('SingletonMeta', (object,), {})): pass

@singleton_object
class AggregateMessages(Singleton):
    @property
    def src(self):
        return sqlfunctions.col(_java_api(SparkContext).SRC())

```
I managed to do this singleton implementation by using parts from: https://michaelgoerz.net/notes/singleton-objects-in-python.html and http://stackoverflow.com/questions/6760685/creating-a-singleton-in-python
I tried that it should work for both python 2 and 3.